File: bindings/rust/fig/src/editor.rs
//! Comment-preserving, in-place editing of a YAML/JSON document.
//!
//! Unlike [`crate::Value::serialize`], which reserializes a whole value,
//! [`Editor`] splices only the bytes of the node you change — comments, key
//! order, blank lines, and quoting style everywhere else are left byte-
//! identical. Inserted values are rendered by fig's serializer (via [`Value`]),
//! then spliced in; the Zig editor re-frames indentation and flow/block context
//! at the site.
//!
//! The value-taking methods come in two forms: [`Editor::replace_value`] &c.
//! take a [`Value`] directly and are always available; the `serde`-gated
//! [`Editor::replace`] &c. accept any `Serialize` type for convenience.

use std::ptr::NonNull;

use crate::error::Error;
use crate::value::{Value, value_text, value_text_with};
use crate::{Format, SerializeOptions, ffi};

/// One step of a path into a document: a mapping key or a sequence index.
#[derive(Clone, Copy, Debug)]
pub enum Segment<'a> {
    Key(&'a str),
    Index(usize),
}

impl<'a> From<&'a str> for Segment<'a> {
    fn from(key: &'a str) -> Self {
        Segment::Key(key)
    }
}

impl From<usize> for Segment<'_> {
    fn from(index: usize) -> Self {
        Segment::Index(index)
    }
}

/// Build the C path array. The returned segments borrow the key bytes of
/// `path`, so the result must not outlive `path` (it never does: it is consumed
/// within a single FFI call).
pub(crate) fn to_ffi_path(path: &[Segment]) -> Vec<ffi::FigPathSegment> {
    path.iter()
        .map(|seg| match seg {
            Segment::Key(k) => ffi::FigPathSegment {
                kind: 0,
                key_ptr: k.as_ptr(),
                key_len: k.len(),
                index: 0,
            },
            Segment::Index(i) => ffi::FigPathSegment {
                kind: 1,
                key_ptr: std::ptr::null(),
                key_len: 0,
                index: *i,
            },
        })
        .collect()
}

/// Build the C `FigStr` array for a key list. The returned entries borrow the
/// key bytes of `keys`, so the result must not outlive `keys` (it never does:
/// it is consumed within a single FFI call).
pub(crate) fn to_ffi_keys<S: AsRef<str>>(keys: &[S]) -> Vec<ffi::FigStr> {
    keys.iter()
        .map(|k| {
            let s = k.as_ref();
            ffi::FigStr {
                ptr: s.as_ptr(),
                len: s.len(),
            }
        })
        .collect()
}

/// An in-place editor over an owned copy of a document's source.
#[derive(Debug)]
pub struct Editor {
    raw: NonNull<ffi::FigEditor>,
    /// The document's format, used to render replacement/insertion splice text in
    /// the same dialect (e.g. a `Value::Str` becomes `"x"` for TOML/JSON but a
    /// bare `x` for YAML). Hardcoding one format would splice syntactically wrong
    /// text for the others and fail the editor's reparse.
    format: Format,
}

impl Editor {
    /// Open an editor over a copy of `input` in the given format.
    pub fn open(input: &[u8], format: Format) -> Result<Self, Error> {
        let mut raw = std::ptr::null_mut();
        let ffi_format: ffi::FigFormat = format.into();
        let status = unsafe {
            ffi::fig_editor_create(input.as_ptr(), input.len(), ffi_format as i32, &mut raw)
        };
        Error::from_status(status)?;
        let raw = NonNull::new(raw).ok_or(Error::Internal)?;
        Ok(Self { raw, format })
    }

    fn ptr(&self) -> *mut ffi::FigEditor {
        self.raw.as_ptr()
    }

    // ── value edits (over `Value`) ──────────────────────────────────────────

    /// Replace the value at `path` with `value` — any `impl Into<Value>`: a
    /// scalar (`9i64`, `"x"`, `true`), a built [`Value`], or a `&Value`.
    pub fn replace_value(&mut self, path: &[Segment], value: impl Into<Value>) -> Result<(), Error> {
        let repl = value_text(&value.into(), self.format)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_replace_val(self.ptr(), p.as_ptr(), p.len(), repl.as_ptr(), repl.len())
        };
        Error::from_status(status)
    }

    /// Replace the key at `path` with `key`.
    pub fn replace_key(&mut self, path: &[Segment], key: &str) -> Result<(), Error> {
        let repl = value_text(&Value::Str(key.to_string()), self.format)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_replace_key(self.ptr(), p.as_ptr(), p.len(), repl.as_ptr(), repl.len())
        };
        Error::from_status(status)
    }

    /// Insert `key: value` into the mapping at `path` (empty path = root).
    /// `value` is any `impl Into<Value>`.
    pub fn insert_value(
        &mut self,
        path: &[Segment],
        key: &str,
        value: impl Into<Value>,
    ) -> Result<(), Error> {
        let key_text = value_text(&Value::Str(key.to_string()), self.format)?;
        let val = value_text(&value.into(), self.format)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_insert_key(
                self.ptr(),
                p.as_ptr(),
                p.len(),
                key_text.as_ptr(),
                key_text.len(),
                val.as_ptr(),
                val.len(),
            )
        };
        Error::from_status(status)
    }

    /// Upsert a mapping value: replace the value at `path`, or insert it when
    /// only the trailing key is absent. Folds the common
    /// `replace_value` → (on [`Error::NotFound`]) `insert_value` two-step into a
    /// single call. `path` must end in a key (it only ever creates a mapping
    /// entry).
    ///
    /// Missing intermediate containers are created for you (`mkdir -p` for
    /// config). For YAML they are created as **block** mappings, so a fresh
    /// nested path reads like hand-written YAML and accepts any value:
    ///
    /// ```text
    /// set_value(a.b.c, 1)  ->  a:
    ///                            b:
    ///                              c: 1
    /// ```
    ///
    /// The dotted-key formats (TOML, fig) instead seed the idiomatic flow chain
    /// (`a = { b = 1 }`), which `fig fmt` canonicalizes to `a.b = 1`.
    ///
    /// Two things to know:
    ///
    /// * A **flow** container cannot hold a block-spelled value. Setting a map or
    ///   sequence *into a flow container the document already has*
    ///   (`a: {b: 1}`, `a: {}`) fails with [`Error::InvalidArgument`] rather than
    ///   writing something that no longer means what was passed in. Expand the
    ///   container to block form first, or pass a value that renders in flow.
    /// * A failed `set_value` is a no-op: any container created on the way is
    ///   rolled back, so `Err` always means the document is unchanged.
    pub fn set_value(&mut self, path: &[Segment], value: impl Into<Value>) -> Result<(), Error> {
        let val = value_text(&value.into(), self.format)?;
        let p = to_ffi_path(path);
        let status =
            unsafe { ffi::fig_editor_set(self.ptr(), p.as_ptr(), p.len(), val.as_ptr(), val.len()) };
        Error::from_status(status)
    }

    // ── value edits with a layout knob (block-vs-inline containers) ─────────
    //
    // The plain `replace_value`/`insert_value`/`set_value` above render a fig
    // container value as inline flow (`k = { … }`) — the only spelling a bare
    // inline splice can carry. These `*_with` twins instead honor `options`
    // (notably `width`): a map/sequence that does not fit renders as a BLOCK
    // section, which the core editor re-frames under the target key
    // (`key` header + `> …` body). This is the width knob for splices — e.g. a
    // short map you want to land one-record-per-line rather than frozen inline.
    // For non-fig formats they behave like the plain methods, plus `options`.

    /// Replace the value at `path`, rendering `value` with `options` (see the
    /// `*_with` note above — a block map/sequence lands as a nested section).
    pub fn replace_value_with(
        &mut self,
        path: &[Segment],
        value: impl Into<Value>,
        options: SerializeOptions,
    ) -> Result<(), Error> {
        let repl = value_text_with(&value.into(), self.format, options)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_replace_val(self.ptr(), p.as_ptr(), p.len(), repl.as_ptr(), repl.len())
        };
        Error::from_status(status)
    }

    /// Insert `key: value` into the mapping at `path`, rendering `value` with
    /// `options` (a block map/sequence lands as a nested section).
    pub fn insert_value_with(
        &mut self,
        path: &[Segment],
        key: &str,
        value: impl Into<Value>,
        options: SerializeOptions,
    ) -> Result<(), Error> {
        let key_text = value_text(&Value::Str(key.to_string()), self.format)?;
        let val = value_text_with(&value.into(), self.format, options)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_insert_key(
                self.ptr(),
                p.as_ptr(),
                p.len(),
                key_text.as_ptr(),
                key_text.len(),
                val.as_ptr(),
                val.len(),
            )
        };
        Error::from_status(status)
    }

    /// Upsert the value at `path`, rendering `value` with `options` (a block
    /// map/sequence lands as a nested section). The width-aware twin of
    /// [`set_value`](Self::set_value).
    pub fn set_value_with(
        &mut self,
        path: &[Segment],
        value: impl Into<Value>,
        options: SerializeOptions,
    ) -> Result<(), Error> {
        let val = value_text_with(&value.into(), self.format, options)?;
        let p = to_ffi_path(path);
        let status =
            unsafe { ffi::fig_editor_set(self.ptr(), p.as_ptr(), p.len(), val.as_ptr(), val.len()) };
        Error::from_status(status)
    }

    /// Append `value` (any `impl Into<Value>`) to the sequence at `path`.
    pub fn append_value(&mut self, path: &[Segment], value: impl Into<Value>) -> Result<(), Error> {
        let val = value_text(&value.into(), self.format)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_append_seq(self.ptr(), p.as_ptr(), p.len(), val.as_ptr(), val.len())
        };
        Error::from_status(status)
    }

    /// Prepend `value` (any `impl Into<Value>`) to the sequence at `path`.
    pub fn prepend_value(&mut self, path: &[Segment], value: impl Into<Value>) -> Result<(), Error> {
        let val = value_text(&value.into(), self.format)?;
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_prepend_seq(self.ptr(), p.as_ptr(), p.len(), val.as_ptr(), val.len())
        };
        Error::from_status(status)
    }

    // ── value edits (serde convenience) ─────────────────────────────────────

    /// Replace the value at `path` with the serialized form of `value`.
    #[cfg(feature = "serde")]
    pub fn replace<T: serde::Serialize + ?Sized>(
        &mut self,
        path: &[Segment],
        value: &T,
    ) -> Result<(), Error> {
        self.replace_value(path, crate::ser::to_value(value)?)
    }

    /// Insert `key: value` into the mapping at `path` (empty path = root).
    #[cfg(feature = "serde")]
    pub fn insert<T: serde::Serialize + ?Sized>(
        &mut self,
        path: &[Segment],
        key: &str,
        value: &T,
    ) -> Result<(), Error> {
        self.insert_value(path, key, crate::ser::to_value(value)?)
    }

    /// Upsert: replace the value at `path`, or insert it when only the trailing
    /// key is absent (see [`set_value`](Self::set_value)).
    #[cfg(feature = "serde")]
    pub fn set<T: serde::Serialize + ?Sized>(
        &mut self,
        path: &[Segment],
        value: &T,
    ) -> Result<(), Error> {
        self.set_value(path, crate::ser::to_value(value)?)
    }

    /// Append the serialized form of `value` to the sequence at `path`.
    #[cfg(feature = "serde")]
    pub fn append<T: serde::Serialize + ?Sized>(
        &mut self,
        path: &[Segment],
        value: &T,
    ) -> Result<(), Error> {
        self.append_value(path, crate::ser::to_value(value)?)
    }

    /// Prepend the serialized form of `value` to the sequence at `path`.
    #[cfg(feature = "serde")]
    pub fn prepend<T: serde::Serialize + ?Sized>(
        &mut self,
        path: &[Segment],
        value: &T,
    ) -> Result<(), Error> {
        self.prepend_value(path, crate::ser::to_value(value)?)
    }

    // ── comment editing ─────────────────────────────────────────────────────

    /// Add an own-line comment ABOVE the node at `path` (the key's line for a
    /// mapping entry), at its indentation, nearest the node. `text` may be
    /// multi-line (one comment line per row). The marker (`#` for YAML, `//` for
    /// JSONC/JSON5) is added for you; strict JSON returns
    /// [`Error::UnsupportedFormat`].
    pub fn add_leading_comment(&mut self, path: &[Segment], text: &str) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_add_leading_comment(self.ptr(), p.as_ptr(), p.len(), text.as_ptr(), text.len())
        };
        Error::from_status(status)
    }

    /// Set the same-line trailing comment on the value at `path`, replacing an
    /// existing one or appending if absent. `text` must be a single line
    /// ([`Error::InvalidArgument`] otherwise); strict JSON returns
    /// [`Error::UnsupportedFormat`].
    pub fn set_trailing_comment(&mut self, path: &[Segment], text: &str) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_set_trailing_comment(self.ptr(), p.as_ptr(), p.len(), text.as_ptr(), text.len())
        };
        Error::from_status(status)
    }

    /// Remove the own-line comment block immediately above the node at `path`.
    /// A no-op (still `Ok`) when there is none.
    pub fn delete_leading_comments(&mut self, path: &[Segment]) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe { ffi::fig_editor_delete_leading_comments(self.ptr(), p.as_ptr(), p.len()) };
        Error::from_status(status)
    }

    /// Remove the same-line trailing comment on the value at `path`. A no-op
    /// (still `Ok`) when there is none.
    pub fn delete_trailing_comment(&mut self, path: &[Segment]) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe { ffi::fig_editor_delete_trailing_comment(self.ptr(), p.as_ptr(), p.len()) };
        Error::from_status(status)
    }

    /// Read the own-line comment block immediately above the node at `path`
    /// (lines joined by `\n`, markers and indentation stripped). `None` when
    /// there is no such block; `Some("")` for a present-but-empty bare marker.
    /// Strict JSON returns [`Error::UnsupportedFormat`].
    pub fn leading_comment(&self, path: &[Segment]) -> Result<Option<String>, Error> {
        self.read_comment(path, false)
    }

    /// Read the same-line trailing comment on the value at `path` (marker
    /// stripped). `None` when there is none; `Some("")` for a bare marker.
    /// Strict JSON returns [`Error::UnsupportedFormat`].
    pub fn trailing_comment(&self, path: &[Segment]) -> Result<Option<String>, Error> {
        self.read_comment(path, true)
    }

    /// Shared body for the two comment reads. Returns an owned `String` (copied
    /// out of the handle's reused scratch buffer) so the result stays valid
    /// across later reads, and maps `NOT_FOUND` to `Ok(None)` — distinguishing an
    /// absent comment from a present-but-empty one (`Some(String::new())`).
    fn read_comment(&self, path: &[Segment], trailing: bool) -> Result<Option<String>, Error> {
        let p = to_ffi_path(path);
        let mut ptr: *const u8 = std::ptr::null();
        let mut len: usize = 0;
        let status = unsafe {
            if trailing {
                ffi::fig_editor_get_trailing_comment(self.ptr(), p.as_ptr(), p.len(), &mut ptr, &mut len)
            } else {
                ffi::fig_editor_get_leading_comment(self.ptr(), p.as_ptr(), p.len(), &mut ptr, &mut len)
            }
        };
        if status.0 == ffi::FigStatus::NOT_FOUND {
            return Ok(None);
        }
        Error::from_status(status)?;
        Ok(Some(borrow_str(ptr, len)?.to_string()))
    }

    // ── structural edits (no value) ─────────────────────────────────────────

    /// Delete the mapping entry named by `path`.
    pub fn delete(&mut self, path: &[Segment]) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe { ffi::fig_editor_delete_key(self.ptr(), p.as_ptr(), p.len()) };
        Error::from_status(status)
    }

    /// Remove the item at `index` from the sequence at `path`.
    pub fn remove_item(&mut self, path: &[Segment], index: usize) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status =
            unsafe { ffi::fig_editor_remove_seq_item(self.ptr(), p.as_ptr(), p.len(), index) };
        Error::from_status(status)
    }

    /// Move the mapping entry at `src_path` to immediately before the entry at
    /// `dest_path`. Both must name keys in the same mapping. The moved entry
    /// keeps its owned comments; bytes between the two entries are preserved.
    pub fn move_key(&mut self, src_path: &[Segment], dest_path: &[Segment]) -> Result<(), Error> {
        let s = to_ffi_path(src_path);
        let d = to_ffi_path(dest_path);
        let status = unsafe {
            ffi::fig_editor_move_key(self.ptr(), s.as_ptr(), s.len(), d.as_ptr(), d.len())
        };
        Error::from_status(status)
    }

    /// Reorder the entries of the mapping at `path` (empty path = root) so
    /// `keys` come first in that order; entries whose key is not listed keep
    /// their original relative order and follow. Unknown keys are ignored. Each
    /// entry's comments and interleaved trivia are preserved.
    pub fn reorder_keys<S: AsRef<str>>(
        &mut self,
        path: &[Segment],
        keys: &[S],
    ) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let k = to_ffi_keys(keys);
        let status = unsafe {
            ffi::fig_editor_reorder_keys(self.ptr(), p.as_ptr(), p.len(), k.as_ptr(), k.len())
        };
        Error::from_status(status)
    }

    /// Move the sequence item at index `from` to index `to` (array-move
    /// semantics). A block item keeps its owned comments; a flow sequence keeps
    /// its separators. No-op when `from == to`.
    pub fn move_item(&mut self, path: &[Segment], from: usize, to: usize) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status =
            unsafe { ffi::fig_editor_move_item(self.ptr(), p.as_ptr(), p.len(), from, to) };
        Error::from_status(status)
    }

    /// Reorder the items of the sequence at `path` so the items at `indices`
    /// (positions in the current order) come first, in that order; items not
    /// listed keep their original relative order and follow. Out-of-range
    /// indices are ignored.
    pub fn reorder_items(&mut self, path: &[Segment], indices: &[usize]) -> Result<(), Error> {
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_reorder_items(
                self.ptr(),
                p.as_ptr(),
                p.len(),
                indices.as_ptr(),
                indices.len(),
            )
        };
        Error::from_status(status)
    }

    /// Reconcile the sequence at `path` so its items are exactly `items`, while
    /// preserving the comments on items that survive the change. Items are
    /// matched to the current ones by value (kind + value, honoring
    /// multiplicity), so a kept or merely reordered item keeps its leading and
    /// trailing comments; only genuinely new values are inserted and only
    /// genuinely dropped values are deleted. The result order matches `items`.
    /// This is the comment-preserving alternative to replacing the whole list.
    ///
    /// Declines with [`Error::InvalidArgument`] (the caller should fall back to
    /// replacing the whole value) when the shape can't be safely diffed: an
    /// empty `items`, an empty current list, a non-scalar item on either side, a
    /// non-sequence target, or a format whose scalars can't stand alone (TOML).
    pub fn set_sequence(&mut self, path: &[Segment], items: &[Value]) -> Result<(), Error> {
        let texts: Vec<String> = items
            .iter()
            .map(|v| value_text(v, self.format))
            .collect::<Result<_, _>>()?;
        let strs = to_ffi_keys(&texts);
        let p = to_ffi_path(path);
        let status = unsafe {
            ffi::fig_editor_set_sequence(self.ptr(), p.as_ptr(), p.len(), strs.as_ptr(), strs.len())
        };
        Error::from_status(status)
    }

    /// The current source. Borrows editor memory; invalidated by the next edit.
    pub fn source(&self) -> Result<&str, Error> {
        let mut ptr: *const u8 = std::ptr::null();
        let mut len: usize = 0;
        let status = unsafe { ffi::fig_editor_source(self.raw.as_ptr(), &mut ptr, &mut len) };
        Error::from_status(status)?;
        borrow_str(ptr, len)
    }
}

impl Drop for Editor {
    fn drop(&mut self) {
        unsafe { ffi::fig_editor_destroy(self.raw.as_ptr()) };
    }
}

/// Interpret `len` bytes at `ptr` (owned by a fig handle, valid for `'a`) as a
/// UTF-8 string slice.
pub(crate) fn borrow_str<'a>(ptr: *const u8, len: usize) -> Result<&'a str, Error> {
    if len == 0 {
        return Ok("");
    }
    // Safety: on success the ABI guarantees `len` bytes at `ptr` owned by the
    // handle and valid until the next mutation / destroy; the caller bounds the
    // returned borrow accordingly.
    let bytes = unsafe { std::slice::from_raw_parts(ptr, len) };
    std::str::from_utf8(bytes).map_err(|_| Error::Utf8)
}

File: bindings/typescript/src/editor.ts
// Comment-preserving, in-place editing of a whole JSON/JSONC/JSON5/YAML/TOML
// document.
//
// Unlike `serialize`, which re-renders a whole value, `Editor` splices only the
// bytes of the node you change — comments, key order, blank lines, and quoting
// everywhere else stay byte-identical. Inserted values are rendered by fig's
// serializer (see `Editable`) and re-framed at the splice site. Release with
// `dispose` (or a `using` declaration).
import { check, FigError, Format, Status } from "./types.ts";
import { encodeKeyList, encodePath, fig, Frame, handleRegistry, readOutSlice } from "./ffi.ts";
import { Editable, type EditFns, type Segment } from "./edit-ops.ts";

const encoder = new TextEncoder();

// Frees the handle of an Editor dropped without dispose() (leak backstop only).
const REGISTRY = handleRegistry((handle) => fig.fig_editor_destroy(handle));

// Editor edits are rendered in the document's own format before splicing (a
// `Value` string becomes `"x"` for TOML/JSON but a bare `x` for YAML); the Zig
// editor then re-frames that text into the document's actual context.
//
// Each entry is a thunk rather than a direct `fig.fig_editor_*` reference: a
// direct reference would read the property off the lazy `fig` proxy at module
// load, forcing wasm instantiation the moment the package is imported — which
// throws on a browser main thread. Deferring the read to call time keeps import
// side-effect-free (see ffi.ts `init`).
const EDITOR_FNS: EditFns = {
  replaceVal: (...a) => fig.fig_editor_replace_val(...a),
  replaceKey: (...a) => fig.fig_editor_replace_key(...a),
  set: (...a) => fig.fig_editor_set(...a),
  insertKey: (...a) => fig.fig_editor_insert_key(...a),
  deleteKey: (...a) => fig.fig_editor_delete_key(...a),
  appendSeq: (...a) => fig.fig_editor_append_seq(...a),
  prependSeq: (...a) => fig.fig_editor_prepend_seq(...a),
  removeSeqItem: (...a) => fig.fig_editor_remove_seq_item(...a),
  moveKey: (...a) => fig.fig_editor_move_key(...a),
  reorderKeys: (...a) => fig.fig_editor_reorder_keys(...a),
  moveItem: (...a) => fig.fig_editor_move_item(...a),
  reorderItems: (...a) => fig.fig_editor_reorder_items(...a),
  setSequence: (...a) => fig.fig_editor_set_sequence(...a),
  addLeadingComment: (...a) => fig.fig_editor_add_leading_comment(...a),
  setTrailingComment: (...a) => fig.fig_editor_set_trailing_comment(...a),
  deleteLeadingComments: (...a) => fig.fig_editor_delete_leading_comments(...a),
  deleteTrailingComment: (...a) => fig.fig_editor_delete_trailing_comment(...a),
  getLeadingComment: (...a) => fig.fig_editor_get_leading_comment(...a),
  getTrailingComment: (...a) => fig.fig_editor_get_trailing_comment(...a),
};

export class Editor extends Editable {
  private constructor(handle: number, format: Format) {
    super(handle, EDITOR_FNS, format);
    REGISTRY?.register(this, handle, this);
  }

  /** Open an editor over a copy of `input` in `format`
   *  (Json/Jsonc/Json5/Yaml/Toml). Empty input is a valid empty document. */
  static open(input: string | Uint8Array, format: Format): Editor {
    const bytes = typeof input === "string" ? encoder.encode(input) : input;
    const frame = new Frame();
    const out = frame.alloc(4);
    try {
      const ptr = frame.bytes(bytes);
      check(fig.fig_editor_create(ptr, bytes.length, format, out), "fig_editor_create");
      const handle = new DataView(fig.memory.buffer).getUint32(out, true);
      if (handle === 0) throw new FigError(Status.InternalError, "fig_editor_create");
      return new Editor(handle, format);
    } finally {
      frame.dispose();
    }
  }

  /** The editor's current source text, reflecting all edits so far. */
  source(): string {
    const frame = new Frame();
    try {
      const scratch = frame.alloc(8);
      check(fig.fig_editor_source(this.live(), scratch, scratch + 4), "fig_editor_source");
      return readOutSlice(scratch);
    } finally {
      frame.dispose();
    }
  }

  // ── whole containers ────────────────────────────────────────────────────
  //
  // The inherited edits address a container the same way they address a
  // scalar: by the one range of source it occupies. A TOML `[header]` table
  // occupies no such range — its body is the lines after the header, and an
  // `[a.b]` header further down the file extends it — and neither does an INI
  // `[section]` or a `fig` block container. At a path naming one, `delete`,
  // `replaceValue`, `moveKey` and `reorderKeys` all fail with
  // `Status.InvalidArgument` rather than rewrite the header and leave its
  // entries behind. These six are the route for those shapes.
  //
  // Support varies by format, and an unsupported op fails with
  // `Status.UnsupportedFormat`: Toml has all six, Ini and Fig the
  // delete/move/reorder three, every other format none. That is not a gap —
  // Yaml, Json and the rest nest a container in one contiguous region, so the
  // inherited edits already handle it.
  //
  // These live on `Editor` rather than on `Editable` because the C ABI has no
  // `fig_embed_*` twins: an embedded region is opened by archetype, and no
  // archetype hosts a scattered-container format today.

  /** Delete the whole container at `path` — every scattered region of its
   *  subtree, leaving interleaved foreign content in place. */
  deleteContainer(path: readonly Segment[]): void {
    const frame = new Frame();
    try {
      const p = encodePath(frame, path);
      check(fig.fig_editor_delete_container(this.live(), p.ptr, p.len), "deleteContainer");
    } finally {
      frame.dispose();
    }
  }

  /** Create a container at `path` whose entries are `body` (verbatim lines in
   *  the document's format, e.g. `ip = "10.0.0.1"\n`), spliced past the
   *  parent's whole subtree so no existing key is reparented. A body that
   *  doesn't parse rolls the document back. */
  insertContainer(path: readonly Segment[], body: string): void {
    const frame = new Frame();
    try {
      const p = encodePath(frame, path);
      const b = frame.str(body);
      check(fig.fig_editor_insert_container(this.live(), p.ptr, p.len, b.ptr, b.len), "insertContainer");
    } finally {
      frame.dispose();
    }
  }

  /** Rename the container at `path` to `newLeaf`, rewriting every line that
   *  names it: renaming `a` rewrites `[a]`, `[a.b]` and `[[a.c]]` alike. */
  renameContainer(path: readonly Segment[], newLeaf: string): void {
    const frame = new Frame();
    try {
      const p = encodePath(frame, path);
      const l = frame.str(newLeaf);
      check(fig.fig_editor_rename_container(this.live(), p.ptr, p.len, l.ptr, l.len), "renameContainer");
    } finally {
      frame.dispose();
    }
  }

  /** Move the container at `src` before the one at `dest`, re-emitting its
   *  scattered fragments contiguously; interleaved foreign containers stay
   *  put. `dest` of `null` moves it to the end of the document. */
  moveContainer(src: readonly Segment[], dest: readonly Segment[] | null): void {
    // The ABI carries "to EOF" as a NULL destination pointer, and an empty path
    // encodes to pointer 0 — so an empty array would silently become EOF here
    // rather than naming the root the way it does everywhere else. The root is
    // not a container and cannot be a destination, so reject it outright
    // instead of quietly meaning something else.
    if (dest !== null && dest.length === 0) {
      throw new FigError(Status.InvalidArgument, "moveContainer: the root is not a valid destination — pass null to move to the end of the document");
    }
    const frame = new Frame();
    try {
      const s = encodePath(frame, src);
      const d = dest === null ? { ptr: 0, len: 0 } : encodePath(frame, dest);
      check(fig.fig_editor_move_container(this.live(), s.ptr, s.len, d.ptr, d.len), "moveContainer");
    } finally {
      frame.dispose();
    }
  }

  /** Reorder top-level containers so those named in `order` come first, in
   *  that order, each re-emitted contiguously at the position the earliest of
   *  them currently occupies. Containers not named keep their places. */
  reorderContainers(order: readonly string[]): void {
    const frame = new Frame();
    try {
      const o = encodeKeyList(frame, order);
      check(fig.fig_editor_reorder_containers(this.live(), o.ptr, o.len), "reorderContainers");
    } finally {
      frame.dispose();
    }
  }

  /** Append an element with body `body` to the container sequence at `path` —
   *  TOML's `[[header]]` array-of-tables append. */
  appendContainerToSeq(path: readonly Segment[], body: string): void {
    const frame = new Frame();
    try {
      const p = encodePath(frame, path);
      const b = frame.str(body);
      check(fig.fig_editor_append_container_to_seq(this.live(), p.ptr, p.len, b.ptr, b.len), "appendContainerToSeq");
    } finally {
      frame.dispose();
    }
  }

  dispose(): void {
    if (this.disposed) return;
    this.disposed = true;
    REGISTRY?.unregister(this);
    fig.fig_editor_destroy(this.handle);
  }
}
